Debug screen for boolean User Defaults

justtwago opened the pull request:
Fixes #22479

This PR introduces a new debug screen for displaying and editing locally stored boolean user defaults. Sometimes it's extremely helpful to change such flags in real-time while testing some features, such as dynamic dashboard cards. The screen has a search feature that filters flags by category/section, key, and site URL.

https://github.com/wordpress-mobile/WordPress-iOS/assets/16563318/197ba38c-303f-4122-9aa9-f7fbb75fbbd1

To test:
- Build a debug app
- Go to Me -> App Settings -> Debug
- [ ] Make sure you can see a new `Boolean User Defaults` section
- Tap on the section and ensure you can see all the stored boolean user defaults
- Smoke test it by editing some flags

## Regression Notes
1. Potential unintended areas of impact
The Debug screen

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
The screen is for devs only

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

justtwago committed:
Create a debug screen for local flags

The screen is read only and displays grouped binary preferences

justtwago committed:
Make binary preferences screen editable

Additionally, refactored the screen to use Dictionaries instead of custom class to simplify the code.

justtwago committed:
Make binary preferences searchable

justtwago committed:
Move binary preferences debug row

Move it to the main section and add an icon

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

justtwago committed:
Rename binary preferences debug screen

Binary preferences renamed to boolean user defaults

justtwago committed:
Display site URLs instead of IDs

Additionally, refactor the code

justtwago committed:
Move booleasn user defaults viewmodel

Move the viewmodel to a separate file and the entire feature to a separate group folder

justtwago committed:
Fix an improper user default saving issue

hassaanelgarem commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,94 @@
+            persistentRepository.set(value, forKey: userDefaultID)
+        } else if var section = allUserDefaultsSections[sectionID] {
+            section[userDefaultID] = BooleanUserDefault(title: userDefaultID, value: value)
+            var sectionValues = section.mapValues { $0.value }
```
```suggestion
            let sectionValues = section.mapValues { $0.value }
```

hassaanelgarem commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,94 @@
+    }
+}
+
+typealias BooleanUserDefaultsSections = [String: BooleanUserDefaults]
+typealias BooleanUserDefaults = [String: BooleanUserDefault]
```
[Suggestion ] `BooleanUserDefaults` and `BooleanUserDefault` have very similar names, which can be a bit confusing. How about finding more distinctive names?

hassaanelgarem replied:
[Nit] I personally prefer having type aliases and structs at the top of the file as it makes the file more readable.

justtwago replied:
Yeah, I do also prefer it on the top, but I noticed that they placed it on the bottom while taking the course and assumed that this is an iOS-specific convention. Happy to hear that it's not :)

hassaanelgarem commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,94 @@
+    }
+
+    private func isSystemUserDefault(_ key: String) -> Bool {
+        key.starts(with: "com.wordpress.")
```
[Question] What does a SystemUserDefault mean? And why are we filtering it out?
I removed the filter, and the defaults we're filtering seem to be related to Gutenberg, not the system.

justtwago replied:
Good catch. I created the method when I tried to filter out some system flags, but it evolved with no changes in the name. Already fixed that.

hassaanelgarem commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,94 @@
+            if let groupedUserDefaults = entryValue as? [String: Bool] {
+                loadedUserDefaultsSections[entryKey] = processGroupedUserDefaults(groupedUserDefaults)
+            } else if let booleanUserDefault = entryValue as? Bool, !isSystemUserDefault(entryKey) {
+                loadedUserDefaultsSections[Strings.otherBooleanUserDefaultsSectionID, default: [:]][entryKey] = BooleanUserDefault(title: entryKey, value: booleanUserDefault)
```
[Suggestion] This line can be simplified to make it more readable by creating an `others` section as a variable outside of the loop. Set the values here, and assign the whole section after the loop. Something like this:
```swift
    func load() {
        let allUserDefaults = persistentRepository.dictionaryRepresentation()
        var loadedUserDefaultsSections = BooleanUserDefaultsSections()
        var otherSection = BooleanUserDefaults()
        for (entryKey, entryValue) in allUserDefaults {
            if let groupedUserDefaults = entryValue as? [String: Bool] {
                loadedUserDefaultsSections[entryKey] = processGroupedUserDefaults(groupedUserDefaults)
            } else if let booleanUserDefault = entryValue as? Bool, !isSystemUserDefault(entryKey) {
                otherSection[entryKey] = BooleanUserDefault(title: entryKey, value: booleanUserDefault)
            }
        }
        loadedUserDefaultsSections[Strings.otherBooleanUserDefaultsSectionID] = otherSection
        allUserDefaultsSections = loadedUserDefaultsSections
    }
```

hassaanelgarem commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,94 @@
+
+    let coreDataStack: CoreDataStack
+
+    private func filterUserDefaults(by query: String) -> BooleanUserDefaultsSections {
```
[Nit]:
```suggestion
    private func filterUserDefaults(by query: String) -> BooleanUserDefaultsSections {
        guard !searchQuery.isEmpty else {
            return allUserDefaultsSections
        }
```
This way `userDefaultsSections` can be simplified to just `filterUserDefaults(by: searchQuery)`

hassaanelgarem reviewed (commented):
Awesome work on your first PR Arty!! 🚀 
I left some comments on the code but it's nothing major.

---
I also have a few general comments:

The toggle buttons don't have consistent colors. Some are light green and others are dark green. It's not a big deal since it's not a user facing feature, but it would be great if we can fix that.
<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/25306722/2db1b533-a57b-4fb3-a7ea-136bccda0e56" width=40%>

---

Finally, I suggest filtering out the feature flags section, as we already have a dedicated debug screen for it.

---

justtwago committed:
Unify user default toggle tint colors

justtwago committed:
Filter out feature flag user defaults

Additionally simplify the code and extract strings to constants

alpavanoglu commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugView.swift:
```diff
@@ -0,0 +1,37 @@
+                                get: { userDefaultsSection[userDefaultKey]?.value ?? false },
+                                set: { newValue in viewModel.updateUserDefault(newValue, forSection: sectionKey, forUserDefault: userDefaultKey) }
+                            )
+                            Toggle(userDefaultsSection[userDefaultKey]!.title, isOn: isOn)
```
I would rather we don't force unwrap a Dictionary. Can we simply pass in an optional in `Toggle`? If not, can we do it the same way as you did above?
`userDefaultsSection[userDefaultKey]?.title ?? ""`

justtwago replied:
Good catch, Alp. Thanks! Already handled it.

justtwago committed:
Improve nil-safety in BooleanUserDefaultsDebugView

justtwago committed:
Add tests to BooleanUserDefaultsDebugViewModel

justtwago commented:
Thanks for all your feedback, folks! I implemented the improvements and added unit tests.

hassaanelgarem commented on WordPress/WordPressTest/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModelTests.swift:
```diff
@@ -0,0 +1,127 @@
+            viewModel.load()
+
+            // Then
+        XCTAssertTrue(viewModel.userDefaultsSections["Other"]?["entry1"]?.value ?? false)
```
This can be a bit more readable:
```suggestion
        let value = try XCTUnwrap(viewModel.userDefaultsSections["Other"]?["entry1"]?.value)
        XCTAssertTrue(value)
```

hassaanelgarem commented on WordPress/WordPressTest/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModelTests.swift:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    func testLoadUserDefaults_WithOtherSection() {
+            // Given
```
Looks like the whole file needs to be realigned.
```suggestion
        // Given
```

hassaanelgarem reviewed (approved):
Thanks for adding unit tests! ❤️ 
I retested, and the PR looks good to me, I just left minor suggestions 🚀

salimbraksa commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,104 @@
+
+    var userDefaultsSections: BooleanUserDefaultsSections {
+        return filterUserDefaults(by: searchQuery)
+    }
```
[Suggestion] Ideally, this should be `@Published`, because it is the list's source of truth, not `allUserDefaultsSections`. The goal is that whenever  `userDefaultsSections` changes, the UI should change as well.

justtwago replied:
In my opinion, `userDefaultsSections` is not the source of truth because it's not a variable that stores data - it's a wrapper over `allUserDefaultsSections` that filters only those sections that should be visible accordingly to the search query. `userDefaultsSections` value cannot be changed because it doesn't store any value but fully depends on `allUserDefaultsSections` and `searchQuery` which _are_ the source of truth.
There is no way to set a value directly to `userDefaultsSections`, we have to change `allUserDefaultsSections`  instead. This is the reason why I used `@Published` for the top level data.

justtwago replied:
Discussed it on Slack and decided that `@Published` has to be used for renderable data only, but for tracking `allUserDefaultsSections` changes we are going to use `didSet` modifier.

salimbraksa commented on WordPress/Classes/ViewRelated/Me/App Settings/Boolean User Defaults/BooleanUserDefaultsDebugViewModel.swift:
```diff
@@ -0,0 +1,104 @@
+            persistentRepository.set(sectionValues, forKey: sectionID)
+        }
+        load()
+    }
```
[Suggestion] I believe that reloading the entire list for every toggle switch might be excessive. Ideally, we should aim to update only the row linked to the user default entry. This could be accomplished by modifying the `BooleanUserDefault` type to:

```swift
final class BooleanUserDefault: ObservableObject {
    let title: String
    @Published var value: Bool

    init(title: String, value: Bool) {
        self.title = title
        self.value = value
    }
}
```

Then you can bind the `value` with the UI.

justtwago replied:
Sounds absolutely reasonable to me! Thanks for catching this, I'll improve the performance of this.

salimbraksa commented:
Hey @justtwago, thanks for wrangling the Debug screen, with unit tests. 🙏 

In my view, the area of the code that could be enhanced is the toggle switching logic. Right now, it reloads the entire list every time a toggle is switched. We can avoid this by updating the way the data is modeled. I’ll provide more details on this tomorrow.

Feel free to merge this PR and we can follow up on this in a separate PR. 👍

justtwago committed:
Replace dictionary-based data to the list-based approach

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/wordpress-mobile/WordPress-iOS/milestone/269">24.2</a>. This milestone is due in less than 4 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

justtwago committed:
Resolve an issue of overriding all rows values in a section

salimbraksa reviewed (approved):
LGTM! @justtwago 🚀
